lib/src/connection_checker.dart

```diff
@@ -0,0 +1,164 @@
+        _checkOverDnsTimeout = checkOverDnsTimeout,
+        _baseUrlLookupAddress = baseUrlLookupAddress,
+        _connectivity = Connectivity(),
+        _connectionTypeMapper = ConnectionTypeMapper(),
+        _hostReachabilityChecker = HostReachabilityChecker(),
+        _initialConnectedValue = Completer(),
+        _connectionTypeStreamController =
+            BehaviorSubject.seeded(ConnectionType.unknown) {
+    _setupInitialConnection().then((_) => _setupConnectivityListener());
+  }
+
+  factory ConnectionChecker({
+    List<InternetAddress>? customCheckAddresses,
+    Duration? customCheckTimeout,
+    String? baseUrlLookupAddress,
+  }) {
+    return _singleton ??= ConnectionChecker._(
+      checkAddresses: customCheckAddresses,
+      checkOverDnsTimeout: customCheckTimeout,
+      baseUrlLookupAddress: baseUrlLookupAddress,
+    );
+  }
+
+  static ConnectionChecker? _singleton;
+
+  final Connectivity _connectivity;
+  final ConnectionTypeMapper _connectionTypeMapper;
+
+  final BehaviorSubject<ConnectionType> _connectionTypeStreamController;
+  late final StreamSubscription<ConnectionType> _connectionTypeSubscription;
+
+  final List<InternetAddress>? _checkAddresses;
+  final Duration? _checkOverDnsTimeout;
+  final String? _baseUrlLookupAddress;
+  final HostReachabilityChecker _hostReachabilityChecker;
+
+  final Completer<bool> _initialConnectedValue;
+
+  Future<void> dispose() async {
+    await _connectionTypeSubscription.cancel();
+    await _connectionTypeStreamController.close();
+  }
+
+  Future<bool> get isConnected async {
+    if (!_initialConnectedValue.isCompleted) {
+      return _initialConnectedValue.future;
+    } else {
+      return _isConnected;
+    }
+  }
+
+  Stream<bool> get connectionStream => CombineLatestStream([
+        _connectionTypeStreamController.stream,
+        Stream.periodic(_connectionRequestInterval),
+      ], (events) => events.first)
+          .flatMap((connectionType) async* {
+            final isHostReachable = await _hostReachable;
+            if (!isHostReachable) {
+              yield false;
+            } else {
+              if (connectionType is ConnectionType &&
+                  connectionType.onlineType) {
+                yield true;
+              } else {
+                yield false;
+              }
+            }
+          })
+          .scan<int>(
+            (attempts, hostReachable, _) {
+              if (!hostReachable) {
+                return attempts + 1;
+              } else {
+                return 0;
+              }
+            },
+            0,
+          )
+          .where((attemps) =>
+              attemps >= _connectionFailureAttempts || attemps == 0)
+          .switchMap((attemps) async* {
+            final hostReachable =
+                attemps >= _connectionFailureAttempts ? false : true;
+
+            if (hostReachable) {
+              final baseUrlReachable = await _baseUrlReachable;
+              yield baseUrlReachable;
```

Comment:
You can also move all this anonymous functions to named functions to make it more readable